src/cljc/matcher_combinators/core.cljc

```diff
@@ -359,7 +359,7 @@
                                      "set or sequential")
                      (validate-input expected
                                      actual
-                                     set?
+                                     (fn [s] (or (set? s) (instance? java.util.Set s)))
```

Comment:
I'm not sure about this solution yet overall, but if we're going to do it, we should include this in the `accept-seq?` case above as well.

Reply:
Maybe we can look to see what seq/set functions in clojure do when given java sets/seqs? If they interoperate, then we could do it here to. If they don't, then maybe there is a specific reason for that.

Reply:
> If they interoperate, then we could do it here to. If they don't, then maybe there is a specific reason for that.

Anecdotally, it looks like set functions like `clojure.set/difference` throw errors on HashSets, e.g.

(require '[clojure.set :as set])

(set/difference #{:a :b} #{:b :c})
;; => #{:a}

(set/difference
 (doto (java.util.HashSet.) (.add :a) (.add :b))
 (doto (java.util.HashSet.) (.add :b) (.add :c)))
;; => 1. Unhandled java.lang.ClassCastException
;;       class java.util.HashSet cannot be cast to class
;;       clojure.lang.IPersistentSet
```

That said, I'd guess that the context for that choice includes considerations that we don't really need to make (general performance and semantic concerns), whereas this is a specific use case.

Reply:
Also:

```clojure
(set/difference
 #{:a :b}
 (doto (java.util.HashSet.) (.add :b) (.add :c)))
;; => #{:a}

(set/difference
 (doto (java.util.HashSet.) (.add :a) (.add :b))
 #{:b :c})
;; => 1. Unhandled java.lang.ClassCastException
```

Reply:
@philomates I think the reason that only instances of `IPersistentSet` return true for `set?` is that `java.util.Set` instances are mutable. For the purposes of matching, since we only ever read them, I think it's fine to treat `java.util.Set` as matchable. WDYT? I've updated the branch to cover the other set-matching cases.